### PR TITLE
[WIP] Unexport 'status' field in K8s.Resource

### PR DESCRIPTION
**Pull request by maqiuyujoyce**
### Change description

Unexported `status` field in K8s.Resource and updated the marshal/unmarshal logic to ensure the unexported `status` field is still passed along.

### Tests you have done

- [N/A] Run `make ready-pr` to ensure this PR is ready for review.
- [X] Perform necessary E2E testing for changed resources.

go test -timeout 1200s -v -tags=integration ./pkg/controller/dynamic/ -test.run TestCreateNoChangeUpdateDelete -run-tests pubsubtopic

### CODE REVIEW

**Comment by justinsb** [comment]
FYI, I don't think there's any reason to abandon this.  Using methods is generally much better at preventing errors than directly accessing fields.

File: pkg/k8s/resource_test.go [hunk 1]
```diff
@@ -83,147 +83,147 @@ func TestResource_IsResourceIDConfigured(t *testing.T) {
 	}
 }
 
-func TestResource_IsSpecOrStatusUpdateRequired(t *testing.T) {
```
Comment (justinsb) [inline pkg/k8s/resource_test.go:86]: What broke here?

File: pkg/krmtotf/resource.go [hunk 2]
```diff
@@ -322,6 +326,21 @@ func (r *Resource) AllTopLevelFieldsAreImmutableOrComputed() bool {
 	return true
 }
 
+func getObservedStateFromStatus(status map[string]interface{}) map[string]interface{} {
+	observedState, exists, _ := unstructured.NestedMap(status, k8s.ObservedStateFieldName)
+	if !exists {
+		observedState = make(map[string]interface{})
```
Comment (justinsb) [inline pkg/krmtotf/resource.go:332]: Probably safer to just return nil, lest someone try to write to it

File: pkg/krmtotf/tftokrm.go [hunk 3]
```diff
@@ -80,26 +80,51 @@ func GetSpecAndStatusFromState(resource *Resource, state *terraform.InstanceStat
...
+	if observedGeneration, ok := resource.GetStatus()["observedGeneration"]; ok {
 		status["observedGeneration"] = deepcopy.DeepCopy(observedGeneration)
 	}
 	if resource.ResourceConfig.ObservedFields != nil {
-		observedState := resolveObservedState(resource, krmState)
-		if len(observedState) > 0 {
-			status["observedState"] = observedState
+		observedFields := resolveObservedFields(resource, krmState)
+		if len(observedFields) > 0 {
+			// Merge the observed fields into the observed state.
```
Comment (justinsb) [inline pkg/krmtotf/tftokrm.go:112]: Should we merge or just build it anew?

File: pkg/krmtotf/fetchlivestate_test.go [hunk 4]
```diff
@@ -30,666 +30,768 @@ import (
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 )
 
-func TestWithFieldsPresetForRead(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (maqiuyujoyce) [reply pkg/krmtotf/fetchlivestate_test.go:33]: Nope, see the answer above.

**Comment by maqiuyujoyce** [comment]
> FYI, I don't think there's any reason to abandon this. Using methods is generally much better at preventing errors than directly accessing fields.

In order to make it work E2E, marshaling/unmarshaling logic for k8s.Resource, krmtotf.Resource, and dcl.Resource need to be updated. This impacts every single resource in admission and reconciliation.

I'm pretty confident about #1353 as the scope of it is quite limited (TF-based resources only, `status` field only, will likely only impact new resources), and we have decent unit tests, mock tests, and integration tests to cover most use cases. But for this PR, the scope of the change is way larger, and I'm quite concerned about merging the change before we have healthy integration tests with good coverage.

I'm proposing we don't include it for 1.115 milestone and only merge the refactored code after we have the existing integration test back to healthy / set up a new integration test flow with good coverage.

File: pkg/krmtotf/tftokrm.go (hunk 3 above)
Comment (maqiuyujoyce) [reply pkg/krmtotf/tftokrm.go]: Just to clarify, what is "build it anew" in this context? Creating a new map?

Observed fields are also part of the observed state but manually configured for `spec` fields in service mappings (#1023, #1059) so I think merging should be the right behavior. Glad to adjust if you are suggesting a simpler implementation.

File: pkg/krmtotf/resource.go (hunk 2 above)
Comment (maqiuyujoyce) [reply pkg/krmtotf/resource.go]: #1430

**Comment by justinsb** [comment]
I propose we get the observedStatus work into 1.116.

Then we can cut the 1.115 branch this week (and give it a few days to stabilize).  Big change there will be AlloyDB, it looks like.

This does presuppose that we 


>marshaling/unmarshaling logic for k8s.Resource, krmtotf.Resource, dcl.Resource, and unstructured.Unstructured need to be updated. 

What does this look like?  Have you tried implementing json.Unmarshaler / Json.Marshaler?  The trick is normally to assemble whatever you want the object to look like (often in a `map[string]any`), and then call json.Marshal on it.

Anyway, I popped this back into the 1.116 milestone, but mostly because I hadn't seen your comments when going through.  I don't mind how we tackle the safety from a code-perspective, we could add tests as I started doing in #1438

**Comment by maqiuyujoyce** [comment]
> What does this look like?

Here is one example: https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/1403/files#diff-565e4cccca460065f11c5e41024f882b8ea7a8cb9eeff6f0e0959b96b75dfd63R414

> Have you tried implementing json.Unmarshaler / Json.Marshaler?

In our codebase, a lot of the marshaling/unmarshaling logic are done by calling util.Marshal() (code here:https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/pkg/util/util.go#L22). I did a search, and found it is called at 45 places, and not every time it is used to unmarshall to a k8s.Resource.

I tried to add special logic to handle k8s.Resource in util.Marshal(), but it gives me a circular dependency issue. Also the raw object is not always a map, but can be any inherited type of k8s.Resource. I consider this task a quick win, so worked around it by manually identifying all the places where k8s.Resource is the target type to unmarshall.

I'm okay with merge the change as is and let our internal integration tests to verify whether the change makes sense.

**Pull request closed without merging**